[Merge] 用户认证系统功能扩展：增加忘记密码、重置密码和个人信息管理功能

FLASHLACK1314 opened the pull request:
## [Merge] 用户认证系统功能扩展：增加忘记密码、重置密码和个人信息管理功能

---

## 变更内容

- **新增**：
    - 增加了忘记密码功能，支持通过邮箱发送重置密码链接
    - 增加了重置密码功能，允许用户通过邮箱验证令牌重置密码
    - 增加了个人信息更新功能，允许用户修改个人资料（用户名、邮箱、电话）
    - 增加了修改密码功能，允许已登录用户修改密码
    - 增加了邮箱验证令牌相关功能，包括创建、验证和管理令牌
- **涉及模块**：
    - 后端控制器（AuthController）
    - 数据访问层（TokenDAO、UserDAO）
    - 业务逻辑层（AuthLogic）
    - 数据传输对象（BackProfileDTO、EmailVerificationTokenDTO、ForgetPasswordResponseDTO）
    - 配置文件和资源（邮件模板、邮箱配置）
- **关键实现**：
    - 实现了邮箱验证令牌的生成和管理机制，使用 Redis 进行分布式事务处理
    - 使用 Hutool 的邮件工具实现邮件发送功能，支持 HTML 邮件模板
    - 实现了密码重置和个人信息管理的业务逻辑，包括输入验证和安全检查
    - 增加了用户资料和密码更新的数据访问方法，确保数据一致性

---

## 测试内容

- 单元测试：
    - 新增了 EmailTokenTest 类，测试邮箱验证令牌的创建、验证、过期和删除等功能
    - 扩展了 UserTest 类，测试用户密码更新、个人资料更新和异常处理
    - 扩展了 AuthTest 类，测试忘记密码、重置密码、个人信息更新和密码修改功能
- 集成测试：
    - 测试了 Redis 缓存与数据库的一致性，特别是在更新用户信息后
    - 测试了邮件发送功能与令牌验证的集成
- 手动测试：
    - 验证了忘记密码流程，包括邮件接收和重置密码页面跳转
    - 验证了个人信息页面修改功能的正常工作
    - 验证了密码修改功能的安全性和有效性

---

## 相关问题

- 问题 1：邮件发送频率限制的实现，防止频繁请求忘记密码
- 问题 2：令牌过期机制的实现，确保安全性和用户体验平衡

---

## 备注

- 添加了邮箱配置文件 mail.setting 到 .gitignore，避免敏感信息泄露
- 优化了密码验证和用户信息更新的业务逻辑，增强了安全性
- 所有新增功能均通过了单元测试和集成测试，确保代码质量

---

## 截止日期

- 无

---

## 请在合并前确保

- [x] 已完成代码审查
- [x] 通过所有测试
- [x] 已更新文档（如有）
- [x] 代码风格符合项目要求
- [x] 已解决所有合并冲突

Commit:
feat(scheduling): 添加邮箱验证令牌功能

- 新增 EmailVerificationTokenDTO 类用于邮箱验证令牌的数据传输
- 在 TokenDAO 中实现创建、验证和删除邮箱验证令牌的方法
- 在 StringConstant 中添加与邮箱验证相关的错误消息常量

Commit:
feat(auth): 添加忘记密码功能- 新增忘记密码接口和相关逻辑
- 添加邮件发送功能- 创建忘记密码响应DTO
- 增加邮件配置文件和模板

Commit:
feat(auth): 实现忘记密码和重置密码功能

- 新增忘记密码接口，支持发送重置密码邮件
- 添加重置密码接口，允许用户通过邮箱令牌重置密码
- 实现邮件发送异步处理，提高系统响应速度
- 增加密码格式校验和邮件发送频率限制- 优化数据库操作，确保数据一致性和安全性

Commit:
feat(auth): 添加用户个人信息更新功能

- 在 AuthController 中添加 /profile 接口，用于更新用户个人信息
- 在 AuthLogic 和 AuthService 中实现用户个人信息的检查和更新逻辑
- 新增 BackProfileDTO 类用于返回更新后的用户信息- 在 UserDAO 中添加 updateUserProfile 方法，用于更新用户档案信息

Commit:
refactor(auth): 优化个人信息更新接口的参数校验逻辑

- 移除了对 name、email 和 phone 参数的非空判断
- 直接使用正则表达式

Commit:
feat(auth): 添加修改用户密码功能

- 在 AuthController 中添加 changePassword 方法，处理修改密码请求- 在 AuthService 接口中定义 changePassword 方法，用于修改密码
- 在 AuthLogic 类中实现 changePassword 方法，包含密码验证和更新逻辑
- 新增密码格式校验和当前密码验证功能
- 使用 PasswordUtil 工具类对新密码进行加密处理

Commit:
test(scheduling): 添加邮箱验证令牌相关方法的单元测试

- 新增 EmailTokenTest 类，包含以下测试方法：
  - testCreateEmailToken: 测试创建邮箱验证令牌
  - testVerifyEmailToken: 测试验证邮箱令牌的有效性  - testExpiredToken: 测试令牌过期逻辑
  - testNonExpiredToken: 测试令牌未过期的情况
  - testDeleteEmailToken: 测试删除邮箱验证令牌
  - testGetEmailTokenCreatAt: 测试获取邮件验证令牌的创建时间
-每个测试方法都包含了相应的断言，以验证功能的正确性
- 测试结束后会清理 Redis 中的测试数据，确保不影响其他测试

Commit:
refactor(user): 优化用户资料更新逻辑并添加单元测试

- 修改 updateUserProfile 方法，增加旧用户对象参数以优化缓存删除逻辑
- 添加 updateUserPassword 方法的单元测试，验证密码更新和缓存清除功能
- 添加 updateUserProfile 方法的单元测试，验证资料更新和缓存清除功能
-增加异常处理单元测试，确保方法在输入为 null 时能正确抛出异常

Commit:
fix(scheduling): 删除过期邮件验证token

- 在验证邮件token过期时，增加删除对应Redis键的操作
- 优化了资源使用，避免过期token占用内存

Commit:
refactor(scheduling): 优化 resetPassword 方法的参数注解

- 在 resetPassword 方法的 userDO 参数上添加 @NotNull 注解
- 这个改动可以增强代码的可读性和健壮性，确保在调用方法时不会传入空值

Commit:
refactor(auth): 优化用户信息更新逻辑

- 移除了用户信息更新后的空值检查
- 使用 assert 语句替代了异常抛出，简化了代码结构

Commit:
fix(auth): 优化密码修改功能

- 在 changePassword 方法中增加了对当前密码、新密码和确认密码的非空校验
- 修复了仅检查新密码是否为空的问题

Commit:
test(auth): 添加用户密码重置和资料更新相关测试用例- 新增忘记密码功能测试用例，包括正常情况和异常情况
- 添加重置密码功能测试用例，验证密码重置流程
- 实现个人信息更新功能的测试用例，包括正常更新和错误情况处理
- 增加修改密码功能的测试用例，验证密码修改逻辑

Commit:
Merge branch 'develop' into feature/fp-cp-up-fla

# Conflicts:
#	src/main/java/com/frontleaves/scheduling/constants/StringConstant.java

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/AuthLogic.java:
```diff
@@ -325,4 +406,161 @@ public UserLoginDTO checkLoginForNewUser(@NotNull UserLoginVO userLoginVO, HttpS
+                    }
+                });
+        // 更新用户姓名
+        userDO.setName(name);
```
不过已经使用了 `Optional` 你这个可以稍微改一下，参考如下

```java
Optional.ofNullable(name)
              .filter(name ->name.matchs(StringConstant.Regular.USER_NAME_REGULAR_EXPRESSION))
              .ifPresentOrElse(userDO::setName, () -> throw new BusinessException("用户名格式错误",  ErrorCode.BODY_ERROR));
```

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/daos/TokenDAO.java:
```diff
@@ -249,4 +247,134 @@ public UserDO getTokenUser(String userToken) throws BusinessException {
+     * @param userDO 用户实体对象
+     * @return 验证令牌的创建时间，如果没有找到则返回0
+     */
+    public long getEmailTokenCreatAt(UserDO userDO) {
```
变量命名

FLASHLACK1314 replied:
这什么意思怎么改

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/AuthLogic.java:
```diff
@@ -325,4 +406,161 @@ public UserLoginDTO checkLoginForNewUser(@NotNull UserLoginVO userLoginVO, HttpS
+        // 使用正则表达式校验密码
+        if (!newPassword.matches(StringConstant.Regular.PASSWORD_REGULAR_EXPRESSION)) {
+            throw new BusinessException("密码格式错误", ErrorCode.BODY_ERROR);
+        }
```
合法验证在 Controller

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/controllers/AuthController.java:
```diff
@@ -120,4 +120,84 @@ public ResponseEntity<BaseResponse<Void>> userRegistered(
+            HttpServletRequest request
+    ){
+        //检查修改内容是否合规
+        UserDO userDO = authService.checkProfile(name, email, phone,request);
```
合法性验证在 Controller

Commit:
test(dao): 添加行政班和专业列表的更新测试

feat(utils): 新增 Excel 解析工具方法

refactor(student): 重构学生信息导入逻辑

- 移除 StudentLogic 中的 parseExcelStudentList 和 parseExcelToStudents 方法
- 新增 ProjectUtil.parseExcelToRowList 和 StudentLogic.parseExcelToStudentList 方法
- 优化学生信息导入的错误处理和日志记录
- 调整批量导入方法中的异常处理逻辑

Commit:
refactor(auth): 优化重置密码验证方法的代码结构和注释

- 重新组织了方法内部的逻辑结构，增加了对新旧密码相同情况的验证
- 更新了方法注释，明确了参数和返回值的含义，提高了代码的可读性和维护性
- 优化了异常处理，使用更具体的异常信息提高了错误提示的准确性

Commit:
refactor(scheduling): 修正 EmailToken 相关方法命名

- 将 TokenDAO 类中的 getEmailTokenCreatAt 方法重命名为 getEmailTokenCreatedAt
- 相应地更新了 EmailTokenTest 类中的测试方法引用

Commit:
refactor(auth): 重构密码修改功能

- 将密码验证逻辑从 AuthLogic 移动到 AuthController
- 移除 AuthService接口中的 confirmPassword 参数
- 更新相关测试用例

Commit:
refactor(auth): 重构用户资料修改逻辑

- 将用户资料修改相关的格式验证逻辑移至 AuthController 层
- 优化 AuthLogic 中的用户资料修改逻辑，提高代码可读性和维护性- 移除 AuthTest 中冗余的测试用例，简化测试代码

Commit:
refactor(test): 移除用户资料更新异常处理的测试代码- 删除了 UserTest 类中的 testUpdateUserProfileExceptionHandling 方法
-该方法测试了在更新用户资料时传入 null 用户对象的情况- 移除该测试可能是为了简化测试用例或避免不必要的异常处理测试

Commit:
fix(scheduling): 优化电子邮件验证令牌的删除逻辑

- 在验证令牌过期后删除相应的电子邮件映射
- 改进 deleteEmailToken 方法，确保同时删除令牌和对应的电子邮件信息

Commit:
feat(env): 添加项目基础 API 和重置密码链接配置

- 在 application-dev-template.yaml 中添加 project配置项，包括 base-api 和 reset-password 链接
- 在 AuthLogic 中使用新配置生成重置密码的 URL，增加了 HTTPS 协议

Commit:
style(scheduling): 优化生成重置密码 URL 的代码

-移除了生成 URL 时的冗余 "https://" 前缀
- 简化了 URL 的拼接逻辑，提高代码可读性

FLASHLACK1314 commented:
@XiaoLFeng 已经优化完成，单元测试和接口测试都测过了
